pinot-query-runtime/src/main/java/org/apache/pinot/query/mailbox/GrpcMailboxService.java

```diff
@@ -44,16 +48,17 @@
  * </ul>
  */
 public class GrpcMailboxService implements MailboxService<TransferableBlock> {
+  private static final Duration MAILBOX_CACHE_EXPIRY = Duration.ofMinutes(5);
   // channel manager
   private final ChannelManager _channelManager;
   private final String _hostname;
   private final int _mailboxPort;
 
   // maintaining a list of registered mailboxes.
-  private final ConcurrentHashMap<String, ReceivingMailbox<TransferableBlock>> _receivingMailboxMap =
-      new ConcurrentHashMap<>();
-  private final ConcurrentHashMap<String, SendingMailbox<TransferableBlock>> _sendingMailboxMap =
-      new ConcurrentHashMap<>();
+  private final Cache<String, ReceivingMailbox<TransferableBlock>> _receivingMailboxCache =
+      CacheBuilder.newBuilder().expireAfterAccess(MAILBOX_CACHE_EXPIRY.toMinutes(), TimeUnit.MINUTES).build();
+  private final Cache<String, SendingMailbox<TransferableBlock>> _sendingMailboxCache =
+      CacheBuilder.newBuilder().expireAfterAccess(MAILBOX_CACHE_EXPIRY.toMinutes(), TimeUnit.MINUTES).build();
```

Comment:
using expireAfter isn't going to work IMO b/c the access pattern needs to also close the associated opChain indicating they are no longer valid.

Reply:
IMO 
1. there's a very good concept we should apply first: the ownership of a mailbox should be handed over from this ConcurrentHashMap to the OpChain when it is being registered. 
  - with this logic i think expiring the mailbox that are still kept in the concurrent hashmap is good. 

2. If we are hitting OOM it is highly unlikely that it is due to the mailbox concurrent hashmap, but b/c the mailbox it is holding on (GRPC or InMemory is holding on to the cached data in the blocking queue but the OpChain that suppose to consume it has already gone. (e.g.. the object itself is small, but it can hold large data buffer)
  - with this in mind I think it is all the more reason to clarify the ownership issue first. 

CC @61yao and @agavra please share your thoughts as well. I know that operator resource management is a big item. But I was thinking if we have a suitable-sized MVP that we can nailed down and then we can continue to iterate.

Reply:
Yes I think we should probably consider making Operator extend AutoCloseable or a related interface and the anchor point that creates the operator chain should ensure closing all the resources associated with the operators.

For example, in future we may want to do fine grained memory management and even decide how much memory to give an operator to execute under. So when the operator chain is set up, this information is passed down to the operators in the form of some OperatorContext etc. OperatorContext itself should be a AutoCloseable. 

This will also take care of the ownership aspect mentioned in the above comment by @walterddr . So when the operator chain finishes execution and is closed, as long as everyone implemented the AutoCloseable, they will free resources either implicitly (if used in try resource) or explicitly in calls to close().

Reply:
Yeah I agree on the ownership logic and closing the OpChain gracefully. I believe @61yao is working on it?

This PR is not to address the design itself and it rather leaves a code-smell, but the intention was to make the main branch usable since right now we were seeing servers hitting OOMs every few hours.

> using expireAfter isn't going to work IMO b/c the access pattern needs to also close the associated opChain indicating they are no longer valid.

I attached a screenshot in the description. This does seem to have made our internal build more usable. My guess is that the following scenario was happening a lot in our build leading to large leaks:

1. Leaf stage is very costly and takes longer than the timeout
2. MailboxReceive for the receiver of the leaf stages have already exited and hung up.
3. When the code finally reaches an InMemory MailboxSend, the mailbox send ends up populating the the queue with the TransferableBlock ultimately leaking it via the map.

Reply:
discussed offline. using the extended expiry time will almost guarantee to have the opChainSchedulerService marks the problematic opChain in a terminal state instead of relying on the cache cancellation to return the state back to OpChainScheduler 

it is a good temp fix for the time being